File: emergency-centre-typescript/src/composite.ts
import Component from "./component";
import { Level } from "./level";
import Sensor from "./sensor";

export default class Composite implements Component {
  private children: Component[] = [];
  private readonly level: Level;
  private readonly name: string;

  constructor(level: Level, name: string, children: Component[] = []) {
    this.level = level;
    this.name = name;
    this.children = children;
  }

  addChild(child: Component): void {
    if (!this.children.includes(child)) {
      this.children.push(child);
    }
  }

  addChildren(children: Component[]): void {
    children.forEach((child) => this.addChild(child));
  }

  remove(component: Component): void {
    this.children = this.children.filter((child) => child !== component);
  }

  getChidren(): Component[] {
    return this.children;
  }

  getSensors(): Sensor[] {
    const sensors: Sensor[] = [];
    for (let child of this.children) {
      if (child instanceof Sensor) {
        sensors.push(child);
      } else if (child instanceof Composite) {
        sensors.push(...child.getSensors());
      }
    }
    return sensors;
  }

  execute(): void {
    console.log(this.toString());
    this.children.forEach((child) => child.execute());
  }

  toString(): string {
    return `${this.level} ${this.name}`;
  }
}

File: emergency-centre-c#/EmergencyCentre/Composite.cs

namespace EmergencyCentreDotNet.EmergencyCentre;
class Composite(string name, Level level) : IComponent
{
    public string Name { get; private set; } = name;
    public Level Level { get; private set; } = level;

    public List<IComponent> Children { get; private set; } = [];

    public void Add(List<IComponent> components)
    {
        foreach (var component in components)
        {
            Add(component);
        }
    }

    public void Add(IComponent component)
    {
        if (Children.Contains(component))
        {
            return;
        }
        Children.Add(component);
    }

    public void Remove(IComponent component)
    {
        Children.Remove(component);
    }

    public HashSet<Sensor> GetSensors()
    {
        HashSet<Sensor> sensors = [];
        foreach (var child in Children)
        {
            if (child is Sensor s)
            {
                sensors.Add(s);
            }
            else if (child is Composite c)
            {
                sensors.UnionWith(c.GetSensors());
            }
        }
        return sensors;
    }

    public override string ToString()
    {
        return $"Composite{{name='{Name}', level={Level}}}";
    }

    public void Execute()
    {
        Console.WriteLine(this);
        Children.ForEach(component => component.Execute());
    }

    public override bool Equals(object? obj)
    {
        return obj is Composite composite &&
               Name == composite.Name &&
               Level == composite.Level;
    }

    public override int GetHashCode()
    {
        return HashCode.Combine(Name, Level);
    }
}
